hazelcast/src/main/java/com/hazelcast/sql/impl/exec/fetch/Fetch.java

```diff
@@ -0,0 +1,146 @@
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.exec.fetch;
+
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+import com.hazelcast.sql.impl.row.EmptyRowBatch;
+import com.hazelcast.sql.impl.row.ListRowBatch;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.row.RowBatch;
+import com.hazelcast.sql.impl.type.converter.Converters;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * LIMIT/OFFSET core support.
+ */
+public class Fetch {
+
+    private final Expression<?> fetch;
+    private final Expression<?> offset;
+
+    private long fetchValue;
+    private long offsetValue;
+
+    private long offsetApplied;
+    private long fetchApplied;
+
+    public Fetch(Expression<?> fetch, Expression<?> offset) {
+        this.fetch = fetch;
+        this.offset = offset;
+    }
+
+    public void setup(ExpressionEvalContext context) {
+        fetchValue = fetch == null ? Long.MAX_VALUE : eval(fetch, true, context);
+        offsetValue = offset == null ? 0 : eval(offset, false, context);
+    }
+
+    public RowBatch apply(RowBatch batch) {
+        if (fetchApplied == fetchValue) {
+            return EmptyRowBatch.INSTANCE;
+        }
+
+        long toOffset = offsetValue - offsetApplied;
+
+        if (toOffset >= batch.getRowCount()) {
+            // Skip the whole batch
+            offsetApplied += batch.getRowCount();
+
+            return EmptyRowBatch.INSTANCE;
+        }
+
+        // Mark offset as finished and adjust the start index accordingly.
+        int startIndex;
+
+        if (toOffset != 0L) {
+            offsetApplied = offsetValue;
+
+            // toOffset is smaller than the batch size and casting to int is safe
+            startIndex = (int) toOffset;
+        } else {
+            startIndex = 0;
+        }
+
+        // Calculate the end position.
+        long toFetch = fetchValue - fetchApplied;
+        long canFetch = batch.getRowCount() - startIndex;
+
+        if (toFetch >= canFetch) {
+            fetchApplied += canFetch;
+
+            // Return the whole remainder.
+            if (startIndex == 0) {
+                return batch;
+            } else {
+                return slice(batch, startIndex, (int) canFetch);
+            }
+        } else {
+            fetchApplied += toFetch;
+
+            return slice(batch, startIndex, (int) toFetch);
+        }
+    }
+
+    public boolean isDone() {
+        return fetchApplied == fetchValue;
+    }
+
+    private long eval(Expression<?> expression, boolean fetch, ExpressionEvalContext context) {
+        assert expression != null;
+
+        String name = fetch ? "LIMIT" : "OFFSET";
+
+        Object val = expression.eval(FetchRow.INSTANCE, context);
+
+        if (val == null) {
+            throw QueryException.error("Value of " + name + " cannot be null");
```

Comment:
The coverage report suggests that `val == null` is not tested.

Reply:
Added more tests in https://github.com/hazelcast/hazelcast/pull/18091/commits/279d6b08d3cf0269aacdb985dd8744750384b584